Uncaptured variable

### Describe the bug

<title>

### Reproduction Steps

```
fn f(g fn (int) string) {
        println('function returned: ${g(42)}')
}
fn main() {
        s := 'hello world'
        f(|i| s)
}
```

### Expected Behavior

Succesful compilation, and output:
```
function returned: hello world
```

### Current Behavior

```shell
~/box/v $ v run issue1618131123.v
==================
/data/data/com.termux/files/usr/tmp/v_10272/issue1618131123.17494216555533418087.tmp.c:2102:9: error: use of undeclared identifier 's'
 2102 |         return s;
      |                ^
/data/data/com.termux/files/usr/tmp/v_10272/issue1618131123.17494216555533418087.tmp.c:7371:26: warning: incompatible pointer types passing 'usize *' (aka 'unsigned int *') to parameter of type 'GC_word *' (aka 'unsigned long *') [-Wincompatible-pointer-types]
 7371 |                 GC_get_heap_usage_safe(&res.heap_size, &res.free_bytes, &res.unmapped_bytes, &res.bytes_since_gc, &res.total_bytes);
      |                                        ^~~~~~~~~~~~~~
/data/data/com.termux/files/usr/include/gc/gc.h:768:70: note: passing argument to parameter here
  768 | GC_API void GC_CALL GC_get_heap_usage_safe(GC_word * /* pheap_size */,
      |                                                                      ^
/data/data/com.termux/files/usr/tmp/v_10272/issue1618131123.17494216555533418087.tmp.c:7371:42: warning: incompatible pointer types passing 'usize *' (aka 'unsigned int *') to parameter of type 'GC_word *' (aka 'unsigned long *') [-Wincompatible-pointer-types]
 7371 |                 GC_get_heap_usage_safe(&res.heap_size, &res.free_bytes, &res.unmapped_bytes, &res.bytes_since_gc, &res.total_bytes);
      |                                                        ^~~~~~~~~~~~~~~
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

~/box/v $
```

### Possible Solution

Provide lambda closure context and access captured variables through it

### Additional Information/Context

_No response_

### V version

V 0.4.3 5f08d45

### Environment details (OS name and version, etc.)

V full version: V 0.4.3 14afda7.5f08d45
OS: termux, 4.19.188-25463247-abA125FXXU2CVL1, #1 SMP PREEMPT Fri Dec 9 20:15:33 KST 2022
Processor: 8 cpus, 32bit, little endian

getwd: /data/data/com.termux/files/home/box/v
vexe: /data/data/com.termux/files/home/box/v/v/v
vexe mtime: 2023-11-13 14:07:19

vroot: OK, value: /data/data/com.termux/files/home/box/v/v
VMODULES: OK, value: /data/data/com.termux/files/home/.vmodules
VTMP: OK, value: /data/data/com.termux/files/usr/tmp/v_10272

Git version: git version 2.41.0
Git vroot status: weekly.2023.44-124-g5f08d45c
.git/config present: true

CC version: clang version 17.0.4
thirdparty/tcc status: thirdparty-unknown-unknown de82a130
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -79,6 +79,7 @@ pub mut:
 	inside_unsafe              bool // true inside `unsafe {}` blocks
 	inside_const               bool // true inside `const ( ... )` blocks
 	inside_anon_fn             bool // true inside `fn() { ... }()`
+	inside_lambda              bool // true inside `|...| ...`
 	inside_ref_lit             bool // true inside `a := &something`
 	inside_defer               bool // true inside `defer {}` blocks
 	inside_fn_arg              bool // `a`, `b` in `a.f(b)`
@@ -2799,6 +2800,10 @@ pub fn (mut c Checker) expr(mut node ast.Expr) ast.Type {
 			return c.int_lit(mut node)
 		}
 		ast.LambdaExpr {
+			c.inside_lambda = true
+			defer {
+				c.inside_lambda = false
+			}
 			return c.lambda_expr(mut node, c.expected_type)
 		}
 		ast.LockExpr {
@@ -3732,8 +3737,13 @@ fn (mut c Checker) ident(mut node ast.Ident) ast.Type {
 					found_var := c.fn_scope.find_var(node.name)
 
 					if found_var != none {
-						c.error('`${node.name}` must be added to the capture list for the closure to be used inside',
-							node.pos)
+						if c.inside_lambda {
+							// Lambdas don't support capturing variables yet, so that's the only hint.
+							c.error('undefined variable `${node.name}`', node.pos)
+						} else {
+							c.error('`${node.name}` must be added to the capture list for the closure to be used inside',
+								node.pos)
+						}
 						return ast.void_type
 					}
 				}

--- a/vlib/v/checker/tests/lambda_undefined_variables_err.out
+++ b/vlib/v/checker/tests/lambda_undefined_variables_err.out
@@ -1,0 +1,33 @@
+vlib/v/checker/tests/lambda_undefined_variables_err.vv:7:2: warning: unused variable: `s2`
+    5 |     f(|x| s1)
+    6 | 
+    7 |     s2 := 'abc'
+      |     ~~
+    8 |     f(|x| s2)
+    9 | }
+vlib/v/checker/tests/lambda_undefined_variables_err.vv:5:8: error: undefined ident: `s1`
+    3 | 
+    4 | fn main() {
+    5 |     f(|x| s1)
+      |           ~~
+    6 | 
+    7 |     s2 := 'abc'
+vlib/v/checker/tests/lambda_undefined_variables_err.vv:5:4: error: `s1` used as value
+    3 | 
+    4 | fn main() {
+    5 |     f(|x| s1)
+      |       ^
+    6 | 
+    7 |     s2 := 'abc'
+vlib/v/checker/tests/lambda_undefined_variables_err.vv:8:8: error: undefined variable `s2`
+    6 | 
+    7 |     s2 := 'abc'
+    8 |     f(|x| s2)
+      |           ~~
+    9 | }
+vlib/v/checker/tests/lambda_undefined_variables_err.vv:8:4: error: `s2` used as value
+    6 | 
+    7 |     s2 := 'abc'
+    8 |     f(|x| s2)
+      |       ^
+    9 | }

--- a/vlib/v/checker/tests/lambda_undefined_variables_err.vv
+++ b/vlib/v/checker/tests/lambda_undefined_variables_err.vv
@@ -1,0 +1,9 @@
+fn f(g fn (int) string) {
+}
+
+fn main() {
+	f(|x| s1)
+
+	s2 := 'abc'
+	f(|x| s2)
+}

--- a/vlib/v/parser/expr.v
+++ b/vlib/v/parser/expr.v
@@ -852,6 +852,7 @@ fn (mut p Parser) lambda_expr() ?ast.LambdaExpr {
 	defer {
 		p.close_scope()
 	}
+	p.scope.detached_from_parent = true
 
 	mut pos := p.tok.pos()
 	mut params := []ast.Ident{}

